hudi-client/src/main/java/org/apache/hudi/index/bloom/HoodieBloomIndexV2.java

```diff
@@ -0,0 +1,321 @@
+import org.apache.hudi.index.HoodieIndex;
+import org.apache.hudi.io.HoodieBloomRangeInfoHandle;
+import org.apache.hudi.io.HoodieKeyLookupHandle;
+import org.apache.hudi.table.HoodieTable;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.api.java.JavaPairRDD;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+/**
+ * Simplified re-implementation of {@link HoodieBloomIndex} that does not rely on caching, or
+ * incurs the overhead of auto-tuning parallelism.
+ */
+public class HoodieBloomIndexV2<T extends HoodieRecordPayload> extends HoodieIndex<T> {
+
+  private static final Logger LOG = LogManager.getLogger(HoodieBloomIndexV2.class);
+
+  public HoodieBloomIndexV2(HoodieWriteConfig config) {
+    super(config);
+  }
+
+  @Override
+  public JavaRDD<HoodieRecord<T>> tagLocation(JavaRDD<HoodieRecord<T>> recordRDD,
+                                              JavaSparkContext jsc,
+                                              HoodieTable<T> hoodieTable) {
+    return recordRDD
+        .sortBy((record) -> String.format("%s-%s", record.getPartitionPath(), record.getRecordKey()),
+                true, config.getBloomIndexV2Parallelism())
+        .mapPartitions((itr) -> new LazyRangeBloomChecker(itr, hoodieTable)).flatMap(List::iterator)
+        .sortBy(Pair::getRight, true, config.getBloomIndexV2Parallelism())
+        .mapPartitions((itr) -> new LazyKeyChecker(itr, hoodieTable))
+        .filter(Option::isPresent)
+        .map(Option::get);
+  }
+
+  @Override
+  public JavaPairRDD<HoodieKey, Option<Pair<String, String>>> fetchRecordLocation(JavaRDD<HoodieKey> hoodieKeys,
+      JavaSparkContext jsc, HoodieTable<T> hoodieTable) {
+    throw new UnsupportedOperationException("Not yet implemented");
+  }
+
+  @Override
+  public boolean rollbackCommit(String commitTime) {
+    // Nope, don't need to do anything.
+    return true;
+  }
+
+  /**
+   * This is not global, since we depend on the partitionPath to do the lookup.
+   */
+  @Override
+  public boolean isGlobal() {
+    return false;
+  }
+
+  /**
+   * No indexes into log files yet.
+   */
+  @Override
+  public boolean canIndexLogFiles() {
+    return false;
+  }
+
+  /**
+   * Bloom filters are stored, into the same data files.
+   */
+  @Override
+  public boolean isImplicitWithStorage() {
+    return true;
+  }
+
+  @Override
+  public JavaRDD<WriteStatus> updateLocation(JavaRDD<WriteStatus> writeStatusRDD, JavaSparkContext jsc,
+                       HoodieTable<T> hoodieTable) {
+    return writeStatusRDD;
+  }
+
+  /**
+   * Given an iterator of hoodie records, returns a pair of candidate HoodieRecord, FileID pairs.
```

Comment:
can we have more descriptive docs to convey what exactly the class is doing. Something like "Given an iterator of hoodie records, returns a pair of candidate HoodieRecord, FileID pairs, by filtering for ranges and bloom for all records with all fileIds" sort of.

Reply:
Done.